### PR TITLE
refactor(jsonrpc): Add structured errors to query method in ViewClient and RPC

### LINKED ISSUE #3848: View query times out for older block
**Describe the bug**
Running view query on a block ~60 days ago returns "Query has timed out", without providing guidance on what to do next.

**To Reproduce**
```
http post http://rpc.mainnet.near.org method=query params:='{"request_type": "call_function", "block_id": 23308486, "account_id": "staked.poolv1.near", "method_name": "get_total_staked_balance", "args_base64": "e30="}' jsonrpc=2.0 id=123
HTTP/1.1 200 OK
Via: 1.1 google
content-length: 106
content-type: application/json
date: Sun, 24 Jan 2021 13:31:31 GMT

{
    "error": {
        "code": -32000,
        "data": "query has timed out",
        "message": "Server error"
    },
    "id": "123",
    "jsonrpc": "2.0"
}
```

**Expected behavior**
Get result of view call or report that node is not archival and doesn't contain such data.

**Version (please complete the following information):**
- mainnet

### PR DESCRIPTION
**Pull request by khorolets**
* add some errors to `runtime/runtime/state_viewer`
* introduce `QueryError` into chain primitives
* refactor runtime `query` method implementation in `neard`
* refactor `view_client` `handle_query` method
* remove routing for `query` method (resolves #3848)

Still need to implement errors on the RPC and Rosetta side. Work in progress. 

I'm creating this draft PR to ask for a review of error structs and error messages.

### CODE REVIEW

File: chain/jsonrpc-primitives/src/types/query.rs
Comment (frol) [inline chain/jsonrpc-primitives/src/types/query.rs:29]: I suggest we apply the following reasoning:

* Each individual error variant should be self-contained, e.g. `NotSyncedYet`, `DoesNotTrackShard`, and even `VMError` imply a certain context. I suggest we capture the context in the naming:
  * `NoSyncedBlocks` ("There are no fully synchronized blocks on the node yet")
  * `UnavailableShard { requested_shard_id: ShardId }` ("The node does not track the shard #{requested_shard_id} where the requested account resides")
  * `FailedFunctionCall { error_message: String }` - we may consider returning `VMError` structure here since It is what we may want to do in `tx` and `broadcast_tx_*` endpoints anyway, and having a debug-formatted string of the enum is not helpful anyway, so we may just as well just go and dump `VMError`
* I suggest we keep our terminology around "does-not-exist" consistent and always use either "unavailable" or "unknown": `UnknownAccount`, `UnknownAccessKey`
* Contract code can be missing on an account, so it is not part of the "does not exist" crew, but rather just: `NoContractCode { contract_account_id: AccountId }`

File: chain/jsonrpc-primitives/src/types/query.rs
Comment (frol) [inline chain/jsonrpc-primitives/src/types/query.rs:20]:
```suggestion
    #[error("Account ID {0} is invalid")]
```

File: chain/jsonrpc-primitives/src/types/query.rs
Comment (frol) [inline chain/jsonrpc-primitives/src/types/query.rs:51]: Please, preserve the comment we had in the original `jsonrpc/lib.rs` implementation about soft-deprecation

File: chain/jsonrpc-primitives/src/types/query.rs
Comment (frol) [inline chain/jsonrpc-primitives/src/types/query.rs:133]: I recall we had some discussion about adding logging for such unreachable states, so not only we can see the counter, but actually address those based on which case we missed. Do we log it anywhere?

File: chain/jsonrpc/src/lib.rs
Comment (frol) [inline chain/jsonrpc/src/lib.rs:568]:
```suggestion
        let query_response = self.view_client_addr.send(query).await??;
```

File: neard/src/lib.rs
Comment (frol) [inline neard/src/lib.rs:31]: Why?

File: neard/src/runtime/errors.rs
Comment (frol) [inline neard/src/runtime/errors.rs:5]: Do we need to be able to construct `RuntimeQueryError` from the `QueryError`?

```suggestion
pub struct RuntimeQueryError(pub QueryError);
```

I also suggest we rename it to `WrappedQueryError` to better indicate the intention that we just want to have a wrapped type here, and also add a comment explaining that we cannot implement `From<>` trait for the original `QueryError` struct since it is a foreign type.

File: runtime/runtime/src/state_viewer/errors.rs
Comment (frol) [inline runtime/runtime/src/state_viewer/errors.rs:20]: Let's prefer using named arguments inside the enums like:

```rust
    InvalidAccountId { requested_account_id: near_primitives::types::AccountId },
    NoContractCode { contract_account_id: near_primitives::types::AccountId },
```

File: runtime/runtime/src/state_viewer/errors.rs
Comment (frol) [inline runtime/runtime/src/state_viewer/errors.rs:95]: This can be replaced with `#[from]` attribute in the relevant enum variant, I believe

File: runtime/runtime/src/state_viewer/mod.rs
Comment (frol) [inline runtime/runtime/src/state_viewer/mod.rs:102]: Replace this match with `?` (`state_update.iter(&prefix)?.iter().map(...)`)

File: chain/client-primitives/src/types.rs
Comment (frol) [inline chain/client-primitives/src/types.rs:287]: just a reminder to align these accordingly to the comment I left to the enum in jsonrpc-primitives

**Review by frol — APPROVED** [review]
We are on the right track!

**Comment by frol** [comment]
@khorolets Another reminder as we discussed last week: Add tests covering the error responses (at least the legal ones, no need to test IOErrors and Unexpected errors); see `neard/tests/rpc_nodes.rs`

File: neard/src/runtime/mod.rs
Comment (frol) [inline neard/src/runtime/mod.rs:1235]: I am not quite sure it is the way we want to wrap things. Can we change the `CallResult(CallResult { result, log })` into 

```rust
struct QueryResponseKind {
   // ...
   CallResult(CallResult)
}

#[serde(tag = "execution_status", rename="SCREAMING_SNAKE_CASE")]
enum CallResult {
    Success { result, logs },
    Failure { error, logs },
}
```

I see it that the JSON RPC response for a failing view function call will be:

```json
{
  "jsonrpc": "2.0",
  "result": {
    "execution_status": "FAILURE",
    "error": ...,
    "logs": ...
  }
}
```

It does not make sense to me to have a failing CallResult as part of the QueryError since it is not the query failed, but the contract code.

**Review by frol — APPROVED** [review]
Just a few nits, and we are ready to merge it!

**Comment by khorolets** [comment]
Backward compatible messages from jsonrpc needs to be done to finish this PR

File: chain/client/src/view_client.rs
Comment (frol) [inline chain/client/src/view_client.rs:244]: I wonder if it is the only place where we need this `RuntimeQueryError`. If so, it might be easier to move the `match` that is currently in the `from` implementation right here inline and avoid the `RuntimeQueryError`. This will reduce the indirection, but will bloat this implementation :disappointed: 

WDYT?

File: chain/client/src/view_client.rs
Comment (frol) [inline chain/client/src/view_client.rs:430]: 0. Consider removing this wrapper (see my other comment in the `view_client.rs`)
1. If we keep it, let's document it
2. If we keep it, let's rename the struct to be more explicit

```suggestion
/// This is a helper struct which is used to implement `From<near_chain_primitives::error::QueryError>`
/// for `near_client_primitives::types::QueryError` (both are foreign types) to be able to use `?` operator.
#[derive(thiserror::Error, Debug)]
#[error(transparent)]
struct WrappedChainQueryError(#[from] pub near_chain::near_chain_primitives::error::QueryError);
```

File: chain/chain-primitives/src/error.rs
Comment (frol) [inline chain/chain-primitives/src/error.rs:30]: I feel like it should be merged into the `InternalError` variant. I feel uneasy about merging a perfectly structured variant into a string, but it seems that after all the `StorageError` is implemented as an internal error representation anyway.

File: chain/jsonrpc-primitives/src/types/receipts.rs
Comment (frol) [inline chain/jsonrpc-primitives/src/types/receipts.rs:57]: I prefer thinking of the logging levels in the following way:

* error - it is a full-stop, there is a need for manual recovery, we should not progress further
* warning - we want to bring the information to the attention, but the program can keep going
* info - whatever we think a user may want to know during the normal execution
* debug - more details on what is happening
* trace - the level at which we want to know every single meaningful step the program makes

Thus, let's make them `warn` instead of `error`

File: neard/tests/rpc_nodes.rs
Comment (frol) [inline neard/tests/rpc_nodes.rs:765]: todo for @frol: This is blocked on #3901

File: runtime/runtime/src/state_viewer/errors.rs
Comment (frol) [inline runtime/runtime/src/state_viewer/errors.rs:24]: I would move `StorageError` to be the last or even consider immediately convert it to `InternalError { error_message: String }` (or `InternalError { #[from] storage_error: StorageError }` wherever possible, and only `ViewAccessKeyError` will have it as a `String`). I don't have a strong preference here. Either way, I would put it the last variant on the list, and definitely name it as an `InternalError`, as from the external user point of view, that is the only thing they care about.

**Review by frol — CHANGES_REQUESTED** [review]

**Comment by khorolets** [comment]
@frol I hope I've finalized error structs according to your suggestions. I'm going to deal with backward now.

File: neard/src/runtime/errors.rs
Comment (frol) [inline neard/src/runtime/errors.rs:5]: Let's just use it by absolute path since it is the only place, and we don't need to import it. It also follows the style we already have

```suggestion
#[easy_ext::ext(FromStateViewerErrors)]
```

**Review by frol — CHANGES_REQUESTED** [review]

File: chain/jsonrpc/src/lib.rs
Comment (frol) [inline chain/jsonrpc/src/lib.rs:580]: :cry: 

I think our only option going forward would be to return both `result` and `error` in the same JSON RPC response. (`result` for backward compatibility, and `error` for future sanity)

File: chain/jsonrpc-primitives/src/types/query.rs
Comment (frol) [inline chain/jsonrpc-primitives/src/types/query.rs:69]: I feel that the nesting is completely unnecessary here. We should either (1) implement our own `RpcQueryResponse` or (2) just type-alias the `views::QueryResponse`. For clear separation of concerns and following our previous style, we should consider (1) and also removing the serde serialization from `view::QueryResponse`.

**Review by frol — CHANGES_REQUESTED** [review]
It is a massive effort! Given there is only one comment from me, I think it is time to mark it is ready for further review.

File: chain/client/src/view_client.rs
Comment (bowenwang1996) [inline chain/client/src/view_client.rs:224]: I am sorry but please do not try to change the logic here. What you did is not an equivalent transformation.

File: chain/jsonrpc/src/lib.rs
Comment (bowenwang1996) [inline chain/jsonrpc/src/lib.rs:244]: what is this change for

File: chain/jsonrpc/src/lib.rs
Comment (bowenwang1996) [reply chain/jsonrpc/src/lib.rs:580]: why couldn't we get rid of `QueryResponseKind::Error` and emulate it here instead?

File: chain/client/src/view_client.rs
Comment (khorolets) [reply chain/client/src/view_client.rs:224]: @bowenwang1996 can you assist here to transform it equivalent, please?

File: chain/jsonrpc/src/lib.rs
Comment (khorolets) [reply chain/jsonrpc/src/lib.rs:244]: Alphabetical ordering of methods, we found it to be easier to read and maintain

File: chain/jsonrpc/src/lib.rs
Comment (frol) [reply chain/jsonrpc/src/lib.rs:580]: @bowenwang1996 good point.

@khorolets Can we get rid of the `Error` variant and move this handling to a separate backward-compatibility-wrapper called on the `process_request` level? (`serde_json::to_value(query_response)` can moved into this wrapper)

File: neard/tests/rpc_nodes.rs
Comment (frol) [inline neard/tests/rpc_nodes.rs:737]: hint: see #4013 on how to use `run_actix_until_stop`  (after the PR is merged)

File: chain/client/src/view_client.rs
Comment (frol) [reply chain/client/src/view_client.rs:224]: @bowenwang1996 we removed query-routing here following your take in https://github.com/near/nearcore/pull/3204/files#diff-34fd63bb882408a6cd040297ab48d277c848f434f9f72da6696aa17e86099405R212-R230

**Review by frol — APPROVED** [review]

File: core/primitives/src/views.rs
Comment (frol) [inline core/primitives/src/views.rs:255]: Clean it up

File: neard/tests/rpc_nodes.rs
Comment (frol) [inline neard/tests/rpc_nodes.rs:837]: Let's do strict matching, ensuring that we don't break the error message even slightly

**Review by frol — APPROVED** [review]

**Review by bowenwang1996 — APPROVED** [review]

**Pull request merged**
